### PR TITLE
Refactor embedded cache

### LINKED ISSUE #14: Refactor to a factory, embed cache
I recommend the following refactoring:

```js
const buildJwksFetch = require('jwks-fetch')

const jwksFetch = buildJwks({
  cache: {
    max: 100, // max number of items in the cache
    ttl: 60 * 1000 // 1 minute
  }
})

cosnt secret = await jwksFetch({
  domain: 'https://exampe.com/',
  alg: 'token_alg',
  kid: 'token_kid'
})

jwksFetch.clear() / / reset the cache
```

I would recommend using https://www.npmjs.com/package/tiny-lru instead of node-cache.

### PR DESCRIPTION
**Pull request by conor909**
Closes #14 

Breaking change

jwksFetch is created with an optional cache constructor, including `max` & `ttl` and returns 3 properties:

`jwksFetch.getSecret`
`jwksFetch.clearCache`
`jwksFetch.cache`  (the whole cache is only returned for use in tests, the proposal is just to expose a clearCache function)

### CODE REVIEW

File: src/jwks-fetch.js [hunk 1]
```diff
@@ -1,54 +1,67 @@
...
-const MISSING_KEY_ERROR = 'No matching key found in the set.'
-const NO_KEYS_ERROR = 'No keys found in the set.'
+const lru = require('tiny-lru')
 
-async function jwksFetch ({ domain, alg, kid, cache = null }) {
-  const cacheKey = `${alg}:${kid}:${domain}`
+const MISSING_KEY_ERROR = 'No matching key found in the set.';
 
-  const cached = cache ? cache.get(cacheKey) : null
+async function buildJwksFetch (cacheProps = { max: 100, ttl: 60 * 1000 }) {
```
Comment (mcollina) [inline src/jwks-fetch.js:8]: why is this an async function?

Note that the default values do not work the way you think they do, e.g. `buildJwksFetch({})` will have `max` as `undefined`.

**Review by mcollina — CHANGES_REQUESTED** [review]

File: src/jwks-fetch.js (hunk 1 above)
Comment (conor909) [reply src/jwks-fetch.js:8]: Ah I see, thanks!

**Review by mcollina — CHANGES_REQUESTED** [review]
Please update the README as well

File: README.md [hunk 2]
```diff
@@ -15,45 +15,49 @@ npm install jwks-fetch
...
+When creating the cache contructor you pass some optional parameters based off the tiny-lru package
+- `max`: Max items to hold in cache, the default setting is 100.
+- `ttl`: Milliseconds an item will remain in cache; lazy expiration upon next get() of an item, the default setting is 60000.
 
-If no match is found, `jwks-fetch` will try to fetch the JWKS from the given `domain`.
+```javascript
+const buildJwksFetch = require('jwks-fetch')
+
+
+const jwksFetch = await buildJwksFetch({
```
Comment (mcollina) [inline README.md:57]:
```suggestion
const jwksFetch = buildJwksFetch({
```

Comment (conor909) [reply README.md]: Yep good spot, thanks

File: src/jwks-fetch.js [hunk 3]
```diff
@@ -1,54 +1,67 @@
...
-    throw error
-  }
+      throw error
+    }
 
-  if (!body.keys || body.keys.length === 0) {
-    throw new Error(NO_KEYS_ERROR)
-  }
+    if (!body.keys || body.keys.length === 0) {
+      throw new Error('No keys found in the set.')
```
Comment (simoneb) [inline src/jwks-fetch.js:38]: extract in a constant as it was before instead of inlining this

File: src/jwks-fetch.js [hunk 4]
```diff
@@ -1,54 +1,67 @@
...
+    return secret
+  }
 
-  // Save the key in the cache
-  cache && cache.set(cacheKey, secret)
-  return secret
+  return {
+    getSecret: getSecret,
+    clearCache: () => cache.clear(),
+    cache
```
Comment (simoneb) [inline src/jwks-fetch.js:63]: if you're returning an object containing the cache, why return a method to clear the cache?

File: src/jwks-fetch.js [hunk 5]
```diff
@@ -1,54 +1,67 @@
...
+    // Save the key in the cache
+    cache.set(cacheKey, secret)
+    return secret
+  }
 
-  // Save the key in the cache
-  cache && cache.set(cacheKey, secret)
-  return secret
+  return {
+    getSecret: getSecret,
```
Comment (simoneb) [inline src/jwks-fetch.js:61]:
```suggestion
    getSecret,
```

File: test/jwks.spec.js [hunk 6]
```diff
@@ -43,38 +44,43 @@ t.test('should fetch the remove jwks and return an error if alg and kid do not m
...
-  const secret = await jwksFetch({ domain: 'https://localhost/', alg: 'RS512', kid: 'KEY' })
+  const jwksFetch = buildJwksFetch()
+  const secret = await jwksFetch.getSecret({ domain: 'https://localhost/', alg: 'RS512', kid: 'KEY' })
   t.ok(secret)
   t.includes(secret, jwks.keys[0].x5c[0])
   t.end()
 })
 
-t.test('if a cache is present and the cached key is null it whould return an error', async t => {
+t.test('if the cached key is null it would return an error', async t => {
```
Comment (simoneb) [inline test/jwks.spec.js:54]:
```suggestion
t.test('if the cached key is null it should return an error', async t => {
```

File: test/jwks.spec.js [hunk 7]
```diff
@@ -138,14 +147,46 @@ t.test('it will throw an error id no keys are found in the JWKS', async t => {
...
+    const jwksFetch = buildJwksFetch()
+    await jwksFetch.getSecret({ domain, alg, kid })
   } catch (e) {
     t.equal(e.message, 'No keys found in the set.')
   }
 
   t.end()
 })
 
+t.test('if initializes without any cache settings it should use default values', async t => {
```
Comment (simoneb) [inline test/jwks.spec.js:159]:
```suggestion
t.test('if initialized without any cache settings it should use default values', async t => {
```

File: src/jwks-fetch.js (hunk 4 above)
Comment (conor909) [reply src/jwks-fetch.js:63]: My thinking was to discourage the user to `get()` or `set()` the cache manually and keep the API & documentation simple.  The proposal was to just expose a way to clear the cache to the user, the whole cache object is only used in testing.

File: src/jwks-fetch.js (hunk 3 above)
Comment (conor909) [reply src/jwks-fetch.js]: yes, thanks!

File: src/jwks-fetch.js (hunk 4 above)
Comment (simoneb) [reply src/jwks-fetch.js:64]: okay, not a big deal

**Review by simoneb — APPROVED** [review]

**Review by mcollina — APPROVED** [review]
lgtm

**Pull request merged**